syntax = "proto3";

message coordinate {
  float x = 1;
  float y = 2;
}

message chunk {
  int32 start_index = 1;
  int32 end_index = 2;
}

message mapInfo {
  chunk indices = 1;
  repeated coordinate centroids = 2;
  string master_id = 3;
  int32 iteration = 4;
}

message keyVal {
  int32 key = 1;
  coordinate value = 2;
}

message keyValDict {
  repeated keyVal dict = 1;
  bool success = 2;
}

message reduceInfo {
  int32 id = 1;
  string master_id = 2;
  int32 iteration = 3;
}

message reply {
  bool success = 1;
}

service KMeans {
  rpc Map(mapInfo) returns (reply) {}
  rpc Reduce(reduceInfo) returns (keyValDict) {}
  rpc GetPartition(reduceInfo) returns (keyValDict) {}
}
